feat!: Stricter rule test validations

DMartens opened the pull request:
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)
Implements part of [RFC 84](https://github.com/eslint/rfcs/pull/84), refs #15104.
As the responding issue is closed, but still planned for v9, I decided to create a draft PR with only the first check implemented.
If this PR is unwanted, free feel to close it.
My plan, if this PR is accepted, is to create one commit per check only for `RuleTester`, then add documentation as some details may change and at last port the changes to `FlatRuleTester`.

<!--
    The following template is intentionally not a markdown checkbox list for the reasons
    explained in https://github.com/eslint/eslint/pull/12848#issuecomment-580302888
-->

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[x] Other, please explain: Rule Tester

#### What changes did you make? (Give an overview)
Assertions added to invalid test case error objects:
- [x] Must contain suggestions if the test case produces suggestions
- [x] Must contain the violation message (message or messageId, but not both)

Assertions added to invalid test case suggestion objects:
- [x] Must contain the suggestion code output
- [x] Must contain the suggestion message (messageId or desc, but not both)
- [x] The suggestion code output must differ from the test case code

Assertions added to all test cases:
- [x] Any autofix output must differ from the test case code
- [x] Cannot have identical test cases (https://github.com/eslint/eslint/pull/17955)
- [x] The optional filename property must be a string
- [x] The optional only property must be a boolean

#### Is there anything you'd like reviewers to focus on?
How should the added suggestion matchers for existing core rule be handled?
For the suggestion existence check there are currently 18 missing suggestions from 2 rules.
Should I put them in the same commit as the check, as a separate commit or fix them all at the end?

<!-- markdownlint-disable-file MD004 -->

bmish commented:
Note that I do have more of this implemented in: https://github.com/bmish/eslint/tree/RFC-84

DMartens commented:
Added the second check for requiring a `message` or `messageId` assuming this PR is welcome.
This made the check for `data` without `messageId` useless as `message` checks there is no `data`, so there has to be a `data` property as then `messageId` is required.
Currently I just fixed up the rule tester test cases and not the test cases for the rules.
How do you want me to structure this PR (e.g. one commit for rule tester and one for updating rule tests), so that you can review it more easily?

bmish commented:
@DMartens good progress on this PR. Are you still working on it? We should try to finish it up soon for ESLint v9.

1. You can pull some of the documentation needed for this PR from my branch https://github.com/eslint/eslint/compare/main...bmish:eslint:RFC-84
2. I would recommend saving the task `Cannot have identical test cases` for a separate PR since it's a much more complex check to get right and shouldn't block the rest of the trivial checks.
3. You shouldn't worry about what commit each change is in.

Let me know if you're stuck on anything.

bmish commented:
Excellent, yes definitely eager to see this PR completed, thanks!

bmish reviewed (commented):
Looking pretty good so far.

I would definitely leave "duplicate test case detection" to a separate PR, as it's much more complicated as described in the [RFC](https://github.com/eslint/rfcs/tree/main/designs/2021-stricter-rule-test-validation#disallow-identical-test-cases), and I personally ran into a lot of edge cases with it and trouble getting it perfect when [implementing](https://github.com/ember-template-lint/ember-template-lint/pull/2279) it for a similar linter.

DMartens commented on docs/src/integrate/nodejs-api.md:
```diff
@@ -822,19 +822,19 @@ A test case is an object with the following properties:
 
     If a string is provided as an error instead of an object, the string is used to assert the `message` of the error.
-* `output` (string, required if the rule fixes code): Asserts the output that will be produced when using this rule for a single pass of autofixing (e.g. with the `--fix` command line flag). If this is `null`, asserts that none of the reported problems suggest autofixes.
+* `output` (string, required if the rule fixes code): Asserts the output that will be produced when using this rule for a single pass of autofixing (e.g. with the `--fix` command line flag). If this is `null`, asserts that none of the reported problems suggest autofixes or the fixer did not create a fix.
```
I added this to clarify as there are two scenarios:
1. The rule only adds fixes for specific messages (so there is no fixer)
2. There is a fixer but it "failed" by returning a nullish value

bmish replied:
That's true.

DMartens commented on tests/lib/rule-tester/rule-tester.js:
```diff
@@ -901,14 +912,28 @@ describe("RuleTester", () => {
+        }, "Error must have either a 'messageId' or 'message'.");
+    });
+
+    it("should throw an error if an error has a property besides message or messageId", () => {
```
This is there to check that the "message" or "messageId" is not checked by checking for an empty object

bmish replied:
Okay, fair enough.

DMartens commented on lib/rule-tester/rule-tester.js:
```diff
@@ -1153,6 +1160,7 @@ class RuleTester {
                     );
                 } else {
                     assert.strictEqual(result.output, item.output, "Output is incorrect.");
+                    assert.notStrictEqual(item.code, item.output, "Use 'output: null' if the rule does not fix this case instead of copying the code.");
```
I think this is a preference (maybe a case for eslint-plugin-eslint-plugin?) as personally I try to be as explicit as possible. As the developer copied the code to output I think they want to keep the output property as there would've been no error forcing them to add the `output` property.

bmish replied:
This doesn't seem to be addressed.

mdjermanovic commented:
> My plan, if this PR is accepted, is to create one commit per check only for `RuleTester`, then add documentation as some details may change and at last port the changes to `FlatRuleTester`.

We're going to remove RuleTester soon, so all changes should be done in FlatRuleTester only.

bmish commented:
Now that the alpha is out, it should be a good time to address these merge conflicts. @DMartens are you able to fix this up in the next week? Want to make sure this is ready for the beta.

DMartens commented:
I should be able to fix those merge issues this week.
Would it be better to start off a new PR as the prior `flat-rule-tester.js` is now named `rule-tester.js` and that is the main part of the PR?

bmish commented:
Great. No need for a new PR. Can just fix the merge conflicts so that the correct file is updated.

bmish commented:
I've started on the duplicate test case detection here:
* https://github.com/eslint/eslint/pull/17955

So we'll need to have both of these PRs ready for the beta in a week.

DMartens commented:
Rebased the PR on the master and added fixes for new changes (two rules had `desc` and `messageId` for suggestions set and an added rule tester test had no `message`).

bmish commented:
Thanks for resolving the conflicts. Time to mark this PR "ready for review"?

bmish reviewed (approved):
I'll approve this. Thanks!

Two minor issues that could be addressed now or later:
1. The wording tweak I commented about.
2. Adding this to the migration guide in the "Stricter `RuleTester` checks" section.

mdjermanovic commented on lib/rule-tester/rule-tester.js:
```diff
@@ -991,6 +992,7 @@ class RuleTester {
                             assert.strictEqual(message.endColumn, error.endColumn, `Error endColumn should be ${error.endColumn}`);
                         }
 
+                        assert.ok(!message.suggestions || hasOwnProperty(error, "suggestions"), `Error should have suggestions on error with message: "${message.message}`);
```
We should also check this for error tests that specify only string/regex.

For example, if we have a test case like this for the `no-array-constructor` rule, it should fail because the error has suggestions:

```js
{
    code: "new Array()",
    errors: ["The array literal notation [] is preferable."]
},
```

mdjermanovic commented on tests/lib/rules/array-callback-return.js:
```diff
@@ -204,7 +204,7 @@ ruleTester.run("array-callback-return", rule, {
         { code: "[\"foo\",\"bar\"].sort(function foo() {})", options: allowImplicitOptions, errors: [{ messageId: "expectedInside", data: { name: "function 'foo'", arrayMethodName: "Array.prototype.sort" } }] },
         { code: "[\"foo\",\"bar\"].toSorted(function foo() {})", options: allowImplicitOptions, errors: [{ messageId: "expectedInside", data: { name: "function 'foo'", arrayMethodName: "Array.prototype.toSorted" } }] },
-        { code: "foo.forEach(x => x)", options: allowImplicitCheckForEach, languageOptions: { ecmaVersion: 6 }, errors: [{ messageId: "expectedNoReturnValue", data: { name: "arrow function", arrayMethodName: "Array.prototype.forEach" } }] },
+        { code: "foo.forEach(x => x)", options: allowImplicitCheckForEach, languageOptions: { parserOptions: { ecmaVersion: 6 } }, errors: [{ messageId: "expectedNoReturnValue", data: { name: "arrow function", arrayMethodName: "Array.prototype.forEach" }, suggestions: [{ messageId: "wrapBraces", output: "foo.forEach(x => {x})" }] }] },
```
Can we keep `languageOptions: { ecmaVersion: 6 }` in this and other updated tests below?

DMartens commented:
Thanks for the code reviews. I should have addressed all raised points in their own commit.
Also I split up the test for the explicit testing of missing suggestions.
Further I have an ideas which I think should be part of this PR (but which are not part of the RFC):
If `data` is not specified for an error or suggestion, then the rule tester should test whether the placeholders are available in the message. For example this would fail:
```javascript
export default {
	messages: {
		id: "{{ expected }}"
	},
	create(context) {
		context.report({ messageId: "id", data: { actual: true });
	},
}
```
This is makes sure that all placeholders can be replaced (especially now that `message` and `messageId` cannot be specified for the same error/suggestion.

mdjermanovic commented:
Looks like one test is failing now.

```
 1) RuleTester
       should throw an error when the expected output is not null and the output does not differ from the code:
     AssertionError: expected [Function] to throw error matching /Use 'output: null' if the rule does …/u but got 'Either omit the \'output\' property o…'
```

mdjermanovic commented:
> Further I have an ideas which I think should be part of this PR (but which are not part of the RFC): If `data` is not specified for an error or suggestion, then the rule tester should test whether the placeholders are available in the message. For example this would be caught when tested with the rule tester:
> 
> ```js
> export default {
> 	messages: {
> 		id: "{{ expected }}"
> 	},
> 	create(context) {
> 		context.report({ messageId: "id", data: { actual: true });
> 	},
> }
> ```
> 
> This is makes sure that all placeholders can be replaced (especially now that `message` and `messageId` cannot be specified for the same error/suggestion.

I think this would need another RFC. Could you please open a separate issue to discuss this?

mdjermanovic commented on docs/src/integrate/nodejs-api.md:
```diff
@@ -784,12 +784,12 @@ Please note that `data` in a test case does not assert `data` passed to `context
-* `output` (string): A code string representing the result of applying the suggestion fix to the input code
+* `desc` (string): The suggestion `desc` value. Must provide this or `messageId`
+* `messageId` (string): The suggestion `messageId` value for suggestions that use `messageId`s. Must provide this or `desc`
+* `data` (object): Placeholder data which can be used in combination with `messageId`. Required if `messageId` is specified and its message uses placeholders
```
```suggestion
* `data` (object): Placeholder data which can be used in combination with `messageId`.
```

mdjermanovic commented on tests/lib/rule-tester/rule-tester.js:
```diff
@@ -2122,13 +2191,27 @@ describe("RuleTester", () => {
                     invalid: [{
                         code: "var foo;",
                         errors: [{
+                            message: "Bad var.",
                             suggestions: [{
                                 messageId: "this-does-not-exist"
                             }]
                         }]
                     }]
                 });
-            }, "Error should have an array of suggestions. Instead received \"undefined\" on error with message: \"Bad var.\"");
+            }, 'Error should have no suggestions on error with message: "Bad var."');
```
This seems wrong. The test case expects suggestions but the actual lint error doesn't have suggestions, so the previous message "Error should have an array of suggestions" was correct.

mdjermanovic commented on tests/lib/rule-tester/rule-tester.js:
```diff
@@ -2107,11 +2175,12 @@ describe("RuleTester", () => {
                         invalid: [{
                             code: "var foo;",
                             errors: [{
+                                message: "Avoid using identifiers named 'foo'.",
                                 suggestions
                             }]
                         }]
                     });
-                }, "Error should have no suggestions on error with message: \"Avoid using identifiers named 'foo'.\"");
+                }, "Error should have suggestions on error with message: \"Avoid using identifiers named 'foo'.\"");
```
This also seems wrong. The test case expects no suggestions but the actual lint message has suggestions, so the previous message "Error should have _no_ suggestions..." was correct.

"Error" here refers to the actual result (lint message produced by Linter). When the actual result doesn't match the expected result, it is assumed that the expected result is correct and the actual isn't (i.e., the test case is correct but the functionality being tested is not working as intended).

mdjermanovic commented on lib/rule-tester/rule-tester.js:
```diff
@@ -1096,6 +1108,7 @@ class RuleTester {
                     );
                 } else {
                     assert.strictEqual(result.output, item.output, "Output is incorrect.");
+                    assert.notStrictEqual(item.code, item.output, "Either omit the 'output' property or set it to null if there is no autofix.");
```
```suggestion
                    assert.notStrictEqual(item.code, item.output, "Test error object 'output' matches 'code'. If no autofix is expected, then omit the 'output' property or set it to null.");
```

I think this error message should first describe the problem it found, which is that 'output' is the same as 'code' in the test case.

mdjermanovic commented on lib/rule-tester/rule-tester.js:
```diff
@@ -1075,8 +1084,11 @@ class RuleTester {
+                                        assert.notStrictEqual(expectedSuggestion.output, item.code, `The output of a suggestion should differ from the original source code for suggestion at index: ${index} on error with message: "${message.message}"`);
+                                    });
+                                } else {
+                                    assert.fail(`Error should have an array of suggestions or a number for the amount of suggestions. Instead received "${util.inspect(message.suggestions)}" on error with message: "${message.message}"`);
```
```suggestion
                                    assert.fail("Test error object property 'suggestions' should be an array or a number");
```

I think this branch relates to invalid `suggestions` in the test case, not to the received actual suggestions on the `message` object.

DMartens commented:
Sorry for the delay. I fixed up the mentioned error messages and added a test case for when the `suggestions` property is neither an array or number which was missing.

mdjermanovic commented:
I think the case described in https://github.com/eslint/eslint/pull/17654#discussion_r1458974783 hasn't been addressed.

mdjermanovic commented on docs/src/integrate/nodejs-api.md:
```diff
@@ -784,12 +784,12 @@ Please note that `data` in a test case does not assert `data` passed to `context
 ### Testing Suggestions
 
-Suggestions can be tested by defining a `suggestions` key on an errors object. The options to check for the suggestions are the following (all are optional):
+Suggestions can be tested by defining a `suggestions` key on an errors object. The options to check for the suggestions are the following:
```
```suggestion
Suggestions can be tested by defining a `suggestions` key on an errors object. If this is a number, it asserts the number of suggestions provided for the error. Otherwise, this should be an array of objects, each containing information about a single provided suggestion. The following properties can be used:
```

To specify that `suggestions` can be a number.

mdjermanovic commented on lib/rule-tester/rule-tester.js:
```diff
@@ -991,81 +992,89 @@ class RuleTester {
-                                if (Array.isArray(message.suggestions) && message.suggestions.length > 0) {
-                                    assert.fail(`Error should have no suggestions on error with message: "${message.message}"`);
-                                }
+                            if (!message.suggestions) {
+                                assert.ok(!error.suggestions || (Array.isArray(error.suggestions) && error.suggestions.length === 0), `Error should have suggestions on error with message: "${message.message}"`);
+                            } else if (!error.suggestions || Array.isArray(error.suggestions) && error.suggestions.length === 0) {
+                                assert.ok(message.suggestions.length === 0, `Error should have no suggestions on error with message: "${message.message}"`);
                             } else {
```
```suggestion
                            if (!error.suggestions || (Array.isArray(error.suggestions) && error.suggestions.length === 0)) {
                                if (message.suggestions) {
                                    assert.fail(`Error should have no suggestions on error with message: "${message.message}"`);
                                }
                            } else {
                                if (!message.suggestions) {
                                    assert.fail(`Error should have suggestions on error with message: "${message.message}"`);
                                }
```

I think we can simplify this to avoid duplicating the `!error.suggestions || (Array.isArray(error.suggestions) && error.suggestions.length === 0` condition.

mdjermanovic commented:
Looks good so far. I think the only remaining thing is to check if the received message has suggestions in this case:

https://github.com/eslint/eslint/blob/10485e8b961d045514bc1e34227cf09867a6c4b7/lib/rule-tester/rule-tester.js#L959-L962

mdjermanovic commented on lib/rule-tester/rule-tester.js:
```diff
@@ -916,6 +920,7 @@ class RuleTester {
 
                         // Just an error message.
                         assertMessageMatches(message.message, error);
+                        assert.ok(message.suggestions === void 0 || message.suggestions.length === 0, `The message has untested. suggestions. Please convert the error at index ${i} into an object by assigning this message to a 'message' property.`);
```
```suggestion
                        assert.ok(message.suggestions === void 0, `Error at index ${i} has suggestions. Please convert the test error into an object and specify 'suggestions' property on it to test suggestions.`);
```

For consistency with the check and the error message in the other case.

mdjermanovic reviewed (approved):
LGTM, thanks!

bmish commented:
Great work on this!
